src/nbl/video/IDescriptorPool.cpp

```diff
@@ -4,60 +4,63 @@
 {
-    const bool allowsFreeingDescriptorSets = m_flags & IDescriptorPool::ECF_FREE_DESCRIPTOR_SET_BIT;
-    if (!allowsFreeingDescriptorSets)
-        return false;
-
-    for (auto i = 0u; i < descriptorSetCount; ++i)
+    const bool allowsFreeing = m_creationParameters.flags.hasFlags(ECF_FREE_DESCRIPTOR_SET_BIT);
+    for (uint32_t t = 0; t < static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT); ++t)
     {
-        for (auto t = 0u; t < asset::EDT_COUNT; ++i)
-        {
-            const auto type = static_cast<asset::E_DESCRIPTOR_TYPE>(t);
+        if (!m_descriptorAllocators[t])
+            continue;
 
-            const uint32_t allocatedOffset = descriptorSets[i]->getDescriptorStorageOffset(type);
-            if (allocatedOffset == ~0u)
-                continue;
+        const uint32_t allocatedCount = m_descriptorAllocators[t]->getAllocatedDescriptorCount(allowsFreeing);
+        std::destroy_n(getDescriptorStorage(static_cast<asset::IDescriptor::E_TYPE>(t)), allocatedCount);
 
-            const uint32_t count = descriptorSets[i]->getLayout()->getTotalDescriptorCount(type);
-            assert(count != 0u);
+        m_descriptorAllocators[t]->reset(allowsFreeing);
+    }
 
-            auto* descriptors = descriptorSets[i]->getAllDescriptors(type);
-            assert(descriptors);
+    m_version.fetch_add(1u);
 
-            for (auto c = 0u; c < count; ++c)
-                descriptors[c].~smart_refctd_ptr();
+    return reset_impl();
+}
 
-            m_generalAllocators[type].free_addr(allocatedOffset, count);
-        }
+IDescriptorPool::SDescriptorOffsets IDescriptorPool::allocateDescriptorOffsets(const IGPUDescriptorSetLayout* layout)
+{
+    SDescriptorOffsets offsets;
 
-        const uint32_t count = descriptorSets[i]->getLayout()->getTotalMutableSamplerCount();
-        if (count > 0)
-        {
-            const uint32_t allocatedOffset = descriptorSets[i]->getMutableSamplerStorageOffset();
-            if (allocatedOffset == ~0u)
-                continue;
+    const bool allowsFreeing = m_creationParameters.flags.hasFlags(ECF_FREE_DESCRIPTOR_SET_BIT);
 
-            auto* samplers = descriptorSets[i]->getAllMutableSamplers();
-            assert(samplers);
+    for (uint32_t i = 0u; i < static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT); ++i)
+    {
+        const auto type = static_cast<asset::IDescriptor::E_TYPE>(i);
+        const auto count = layout->getTotalDescriptorCount(type);
+        if (count == 0ull)
+            continue;
 
-            for (auto c = 0u; c < count; ++c)
-                samplers[c].~smart_refctd_ptr();
+        offsets.data[i] = m_descriptorAllocators[i]->allocate(count, allowsFreeing);
 
-            m_generalAllocators[asset::EDT_COUNT].free_addr(allocatedOffset, count);
-        }
+        assert((offsets.data[i] < m_creationParameters.maxDescriptorCount[i]) && "PANIC: Allocation failed. This shoudn't have happened! Check your descriptor pool.");
```

Comment:
you can't just assert, need to handle failures gracefully and return a false (offsets you need to return in a different way now)

Ideally by "deallocating" the recently allocated offsets from previous type's allocators, if any subsequent allocation fails.

This is trivial for a `GeneralPurposeAddressAllocator`, and for the `LinearAddressAllocator` you can use `reset(oldCursorValue)` which "rewinds" the allocator to its previous value.

Reply:
The `oldCursorValue` just so happens to be your newly allocated offset :)

Reply:
So it would need to look like this
```cpp
offsets = {INVALID, ... ,INVALID};
bool success=true

// try to alloc offsets
{
}

if (!success)
{
   // dealloc any offsets which were successfuly allocated
   {
   }
   return false;
}

// construct the storage
{
}
return true;
```

Reply:
Actually you need a full blown `cancel` function and not handle it in the `allocateDescriptorOffsets`

https://github.com/Devsh-Graphics-Programming/Nabla/pull/345/files#r1118735531

Reply:
Resolved.